internal/lsp/documentsymbol.go

```diff
@@ -0,0 +1,267 @@
+
+	"github.com/styrainc/regal/internal/lsp/types"
+	"github.com/styrainc/regal/internal/lsp/types/symbols"
+)
+
+//nolint:nestif
+func documentSymbols(
+	contents string,
+	module *ast.Module,
+) []types.DocumentSymbol {
+	docSymbols := make([]types.DocumentSymbol, 0)
+	pkgSymbols := make([]types.DocumentSymbol, 0)
+
+	lines := strings.Split(contents, "\n")
+
+	pkgRange := types.Range{
+		Start: types.Position{Line: 0, Character: 0},
+		End:   types.Position{Line: uint(len(lines) - 1), Character: uint(len(lines[len(lines)-1]))},
+	}
+
+	pkg := types.DocumentSymbol{
+		Name:           module.Package.Path.String(),
+		Kind:           symbols.Package,
+		Range:          pkgRange,
+		SelectionRange: pkgRange,
+	}
+
+	// Create groups of rules and functions sharing the same name
+	ruleGroups := make(map[string][]*ast.Rule, len(module.Rules))
+
+	for _, rule := range module.Rules {
+		name := refToString(rule.Head.Ref())
+		ruleGroups[name] = append(ruleGroups[name], rule)
+	}
+
+	for _, rules := range ruleGroups {
+		if len(rules) == 1 {
+			rule := rules[0]
+
+			kind := symbols.Variable
+			if isConstant(rule) {
+				kind = symbols.Constant
+			} else if rule.Head.Args != nil {
+				kind = symbols.Function
+			}
+
+			ruleRange := locationToRange(rule.Location)
+			ruleSymbol := types.DocumentSymbol{
+				Name:           refToString(rule.Head.Ref()),
+				Kind:           kind,
+				Range:          ruleRange,
+				SelectionRange: ruleRange,
+			}
+
+			if detail := getRuleDetail(rule); detail != "" {
+				ruleSymbol.Detail = &detail
+			}
+
+			pkgSymbols = append(pkgSymbols, ruleSymbol)
+		} else {
+			groupFirstRange := locationToRange(rules[0].Location)
+			groupLastRange := locationToRange(rules[len(rules)-1].Location)
+
+			groupRange := types.Range{
+				Start: groupFirstRange.Start,
+				End:   groupLastRange.End,
+			}
+
+			kind := symbols.Variable
+			if rules[0].Head.Args != nil {
+				kind = symbols.Function
+			}
+
+			groupSymbol := types.DocumentSymbol{
+				Name:           refToString(rules[0].Head.Ref()),
+				Kind:           kind,
+				Range:          groupRange,
+				SelectionRange: groupRange,
+			}
+
+			detail := getRuleDetail(rules[0])
+			if detail != "" {
+				groupSymbol.Detail = &detail
+			}
+
+			children := make([]types.DocumentSymbol, 0, len(rules))
+
+			for i, rule := range rules {
+				childRange := locationToRange(rule.Location)
+				childRule := types.DocumentSymbol{
+					Name:           fmt.Sprintf("#%d", i+1),
+					Kind:           kind,
+					Range:          childRange,
+					SelectionRange: childRange,
+				}
+
+				childDetail := getRuleDetail(rule)
+				if childDetail != "" {
+					childRule.Detail = &childDetail
+				}
+
+				children = append(children, childRule)
+			}
+
+			groupSymbol.Children = &children
+
+			pkgSymbols = append(pkgSymbols, groupSymbol)
+		}
+	}
+
+	if len(pkgSymbols) > 0 {
+		pkg.Children = &pkgSymbols
+	}
+
+	docSymbols = append(docSymbols, pkg)
```

Comment:
is `docSymbols` doing something I'm missing? It looks like it'll only ever be the pkgSymbols right?

Reply:
No, you're right. I was probably too cautious, and wanted to prepare for the case where two "packages" was in the same document, like when general ref heads rules are used. But I guess even in those cases one will be a child of the other.

I don't think it hurts having a top-level array of symbols though, should we later want to add anything else there. But I can remove it if you think it's redundant.

Reply:
nah it's fine!

Reply:
Perhaps a comment for the intended purpose of docSymbols would be enough? / helpful

Reply:
Sure, I can add a comment 👍